<a href="https://colab.research.google.com/github/turayusa/homework/blob/main/Dx601Fundermental_of_maths_Hypothesis_Testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd              # Load and organize your diabetes data
import numpy as np               # Calculations, including square roots
import matplotlib.pyplot as plt  # Create graphs
from scipy import stats          # Confidence intervals and statistical tests

In [6]:
from google.colab import files

uploaded = files.upload()

Saving diabetic_data.csv to diabetic_data (2).csv


In [7]:
filename = next(iter(uploaded))
df = pd.read_csv(filename)
# display the first five records
df.head()


,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


In [9]:


# 1A. Calculate sample size
hospital_stay = df["time_in_hospital"].dropna()

n = hospital_stay.count()       # n = number of nonmissing values

print("sample size(n)", n)


sample size(n) 101766


### 1B. Define the hypotheses

**Learning objective:** Write null and alternative hypotheses
for the mean hospital stay.

**Null hypothesis:** The population mean hospital stay is 4 days.

$$
H_0: \mu = 4
$$

**Alternative hypothesis:** The population mean hospital stay
is different from 4 days.

$$
H_1: \mu \ne 4
$$

Here, μ represents the population mean.
This is a two-tailed test because “different” includes
both greater than and less than 4 days.

In [12]:



# 1B. Define the hypothesized population mean
mu_0 = 4  # H₀: μ = 4 days

print("Null hypothesis: μ = ", mu_0, "days")
print("Alternative hypothesis: μ ≠", mu_0, "days")

Null hypothesis: μ =  4 days
Alternative hypothesis: μ ≠ 4 days


### 1C. Calculate the sample mean

**Learning objective:** Calculate the average hospital stay
from the diabetes data as preparation for hypothesis testing.

$$
\bar{x} = \frac{\sum_{i=1}^{n} x_i}{n}
$$

Add all recorded hospital-stay values, then divide by
the number of records.



In [13]:

#  1c. Calculate thsample mean
total_days = hospital_stay.sum()   # Σxᵢ
mean_stay = total_days / n         # x̄ = Σxᵢ / n

print("Total hospital days:", total_days)
print("Sample mean:", mean_stay, "days")
print("Hypothesized population mean:", mu_0, "days")

Total hospital days: 447362
Sample mean: 4.395986871843248 days
Hypothesized population mean: 4 days


### 1D. Calculate sample variance

**Learning objective:** Calculate how much hospital stays vary
around the sample mean, preparing for statistical testing.

$$
s^2 = \frac{\sum_{i=1}^{n}(x_i-\bar{x})^2}{n-1}
$$

1. Subtract the sample mean from each hospital stay.
2. Square each difference.
3. Add the squared differences.
4. Divide by n − 1.

In [17]:

# 1D. Calculate sample variance

deviations = hospital_stay - mean_stay  # xᵢ − x̄
squared_deviations = deviations ** 2   # (xᵢ − x̄)²
sum_squared = squared_deviations.sum() # Σ(xᵢ − x̄)²
variance_stay = sum_squared / (n - 1)   # s² = Σ(xᵢ − x̄)² / (n − 1)

print("Sample variance:", variance_stay, "days squared")

Sample variance: 8.910868383417293 days squared


### 1E. Calculate sample standard deviation

**Learning objective:** Measure the spread of hospital stays
around the sample mean in days.

$$
s = \sqrt{s^2}
  = \sqrt{\frac{\sum_{i=1}^{n}(x_i-\bar{x})^2}{n-1}}
$$

Standard deviation is the square root of sample variance.

In [18]:
# 1E. Calculate sample standard deviation
std_stay = np.sqrt(variance_stay)  # s = √(s²)
print("sample standard deviation", std_stay, "days")

sample standard deviation 2.985107767471267 days


### 1F. Calculate the standard error of the mean

**Learning objective:** Calculate the standard error used
in confidence intervals and hypothesis tests.

$$
SE = \frac{s}{\sqrt{n}}
$$

Divide the sample standard deviation by the square root
of the sample size.

- s: sample standard deviation
- n: number of nonmissing records
- SE: estimated variability of the sample mean

In [19]:
# 1F. Calculate the standard error of the mean

sqrt_n = np.sqrt(n)          # √n
se_stay = std_stay / sqrt_n  # SE = s / √n

print("Sample size:", n)
print("Standard error:", se_stay, "days")

Sample size: 101766
Standard error: 0.009357474714516381 days


### 1G. Calculate the t-statistic

**Learning objective:** Calculate a test statistic to assess
the null hypothesis that the population mean stay is 4 days.

$$
t = \frac{\bar{x}-\mu_0}{SE}
  = \frac{\bar{x}-\mu_0}{s/\sqrt{n}}
$$

Subtract the hypothesized mean from the sample mean,
then divide by the standard error.

In [20]:
# 1G. Calculate the t-statistic

mean_difference = mean_stay - mu_0  # x̄ − μ₀
t_stat = mean_difference / se_stay  # t = (x̄ − μ₀) / SE

print("Mean difference:", mean_difference, "days")
print("t-statistic:", t_stat)

Mean difference: 0.395986871843248 days
t-statistic: 42.317706851929614


### 1H. Calculate the two-tailed p-value

**Learning objective:** Calculate the p-value for testing
whether the population mean hospital stay differs from 4 days.

**Degrees of freedom:**

$$
df = n - 1
$$

**Two-tailed p-value:**

$$
p = 2P(T_{n-1} \ge |t|)
$$

Assuming the null hypothesis and test assumptions hold,
the p-value is the probability of a t-statistic at least
as extreme as ours in either direction.

In [21]:
# 1H. Calculate the two-tailed p-value

degrees_freedom = n - 1  # df = n − 1

p_value = 2 * stats.t.sf(
    abs(t_stat), degrees_freedom
)  # p = 2P(T ≥ |t|)

print("Degrees of freedom:", degrees_freedom)
print("Two-tailed p-value:", p_value)

Degrees of freedom: 101765
Two-tailed p-value: 0.0


### 1I. Make the hypothesis-test decision

**Learning objective:** Interpret the p-value to reject
or fail to reject the null hypothesis.

Use a significance level of 0.05 for this practice.

$$
\alpha = 0.05
$$

$$
p < \alpha \Rightarrow \text{Reject } H_0
$$

$$
p \ge \alpha \Rightarrow \text{Fail to reject } H_0
$$

In [22]:
# 1I. Make the hypothesis-test decision

alpha = 0.05  # Significance level: α = 0.05

if p_value < alpha:  # p < α
    print("Reject H₀.")
    print("Evidence suggests the population mean differs from 4 days.")
else:
    print("Fail to reject H₀.")
    print("Insufficient evidence that the population mean differs from 4 days.")

Reject H₀.
Evidence suggests the population mean differs from 4 days.


### Topic Summary: One-Sample Hypothesis Testing

**Learning objective achieved:** Calculate and interpret
a two-tailed one-sample t-test.

| Measure | Result |
|---|---:|
| Sample size | 101,766 |
| Sample mean | 4.396 days |
| Sample variance | 8.9109 days² |
| Standard deviation | 2.9851 days |
| Standard error | 0.0093575 days |
| t-statistic | 42.3177 |
| Degrees of freedom | 101,765 |
| p-value | < 0.001 |
| Significance level | 0.05 |
| Decision | Reject H₀ |

**Interpretation:** Under the test assumptions, there is
evidence that the population mean hospital stay differs
from 4 days. The observed difference is 0.396 days.

**Limitation:** Repeated admissions from the same patient
may violate independence. Statistical significance alone
does not establish clinical importance.

### 2A. Understand Type I and Type II Errors

**Learning objective:** Explain errors in hypothesis-test decisions.

**Type I error:** Reject a true null hypothesis.

$$
\alpha = P(\text{Reject } H_0 \mid H_0 \text{ is true})
$$

**Type II error:** Fail to reject a false null hypothesis.

$$
\beta = P(\text{Fail to reject } H_0 \mid H_0 \text{ is false})
$$

**Diabetes example:**
- Type I: conclude the mean differs from 4 days when it equals 4.
- Type II: fail to detect that the mean differs from 4 days.

In [23]:
# 2A. Understand Type I and Type II errors
# Learning objective: Explain errors in hypothesis-test decisions

# Type I error
# α = P(Reject H₀ | H₀ is true)
print("Type I error: Reject a true H₀.")
print("Vaccine: Conclude risk is reduced when it is not.")
print("Diabetes: Conclude μ ≠ 4 when the true μ = 4.")

print()  # Add a blank line

# Type II error
# β = P(Fail to reject H₀ | H₀ is false)
print("Type II error: Fail to reject a false H₀.")
print("Vaccine: Fail to detect a real reduction in risk.")
print("Diabetes: Fail to detect μ ≠ 4 when it really differs.")

Type I error: Reject a true H₀.
Vaccine: Conclude risk is reduced when it is not.
Diabetes: Conclude μ ≠ 4 when the true μ = 4.

Type II error: Fail to reject a false H₀.
Vaccine: Fail to detect a real reduction in risk.
Diabetes: Fail to detect μ ≠ 4 when it really differs.


### 2B. Four Possible Outcomes of a Hypothesis Test

For our diabetes example:

$$
H_0: \mu = 4
$$

| Decision | H₀ is true: μ = 4 | H₀ is false: μ ≠ 4 |
|---|---|---|
| Fail to reject H₀ | Correct: true negative (TN) | Type II error: false negative (FN) |
| Reject H₀ | Type I error: false positive (FP) | Correct: true positive (TP) |

**Instructor’s vaccine example:**
- False positive: conclude the vaccine reduces risk when it does not.
- False negative: miss a real reduction in risk.

**Diabetes example:**
- False positive: conclude the mean differs from 4 when it equals 4.
- False negative: miss a real difference from 4.

In [25]:
# 2B. Identify a hypothesis-test outcome
# Learning objective: Identify Type I and Type II errors
# Hypothetical diabetes example: H₀: μ = 4 days

null_is_true = False  # Assume the true mean differs from 4 days
reject_null = False  # Assume the test fails to reject H₀

if reject_null and null_is_true:
    print("Type I error: false positive")
    print("Conclude the mean differs from 4 when it actually equals 4.")

elif not reject_null and not null_is_true:
    print("Type II error: false negative")
    print("Fail to detect that the true mean differs from 4 days.")

elif reject_null and not null_is_true:
    print("Correct decision: true positive")
    print("Detect that the true mean differs from 4 days.")

else:
    print("Correct decision: true negative")
    print("Fail to reject H₀ when the true mean equals 4 days.")

Type II error: false negative
Fail to detect that the true mean differs from 4 days.


### Summary: Type I and Type II Errors

**Learning objective achieved:** Identify errors in hypothesis-test decisions.

| Error | Meaning | Vaccine example | Diabetes example |
|---|---|---|---|
| Type I: false positive | Reject a true H₀ | Claim a risk reduction that does not exist | Claim μ ≠ 4 when μ = 4 |
| Type II: false negative | Fail to reject a false H₀ | Miss a real risk reduction | Miss a real difference from 4 days |

**Coding tools practiced:** True, False, if, elif, else, and, not.

These coding scenarios are hypothetical. They do not establish
whether our actual diabetes test made an error.

### 2C. Calculate Statistical Power

**Learning objective:** Calculate the probability of correctly
rejecting a false null hypothesis.

$$
\text{Power} = 1 - \beta
$$

Here, β is the probability of a Type II error.

**Instructor’s vaccine example:** Power is the probability
of detecting a real reduction in disease risk.

**Diabetes example:** Power is the probability of detecting
a specified real difference from the 4-day benchmark.

For practice, assume β = 0.20:

$$
\text{Power} = 1 - 0.20 = 0.80 = 80\%
$$

In [26]:
# 2C. Calculate statistical power

beta = 0.20           # Assumed Type II error probability
power = 1 - beta      # Power = 1 − β
power_percent = power * 100  # Convert to percent

print("Statistical power:", power)
print("Statistical power (%):", power_percent)

Statistical power: 0.8
Statistical power (%): 80.0


### Summary: Statistical Power

**Learning objective achieved:** Calculate power from β.

$$
\text{Power} = 1 - \beta = 1 - 0.20 = 0.80
$$

- β = 20%: probability of missing a specified real effect.
- Power = 80%: probability of detecting that effect.
- Vaccine example: detect a real reduction in disease risk.
- Diabetes example: detect a specified real difference from 4 days.

**Coding tools:** `-` subtracts; `* 100` converts to percent.

### 3A. Calculate a Right-Tailed p-Value

**Learning objective:** Calculate a p-value for a directional test.

$$
H_0: \mu \le 4
\qquad
H_1: \mu > 4
$$

Calculate the test statistic at the null boundary μ₀ = 4.

$$
p_{\text{right}} = P(T_{n-1} \ge t_{\text{observed}})
$$

Only the right tail is used because the alternative says
“greater than 4 days.”

In [27]:
# 3A. Calculate a right-tailed p-value
# Reuse the t-statistic from our diabetes calculations

p_right = stats.t.sf(t_stat, degrees_freedom)  # P(T ≥ t)

print("t-statistic:", t_stat)
print("Right-tailed p-value:", p_right)

t-statistic: 42.317706851929614
Right-tailed p-value: 0.0


### 3B. Calculate a Left-Tailed p-Value

**Learning objective:** Calculate a p-value for a “less than” test.

$$
H_0: \mu \ge 4
\qquad
H_1: \mu < 4
$$

Calculate at the null boundary μ₀ = 4.

$$
p_{\text{left}} = P(T_{n-1} \le t_{\text{observed}})
$$

**Instructor’s directional example:** Test whether one model’s
error is lower than another’s.

**Diabetes example:** Test whether the mean stay is below 4 days.

In [28]:
# 3B. Calculate a left-tailed p-value

p_left = stats.t.cdf(t_stat, degrees_freedom)  # P(T ≤ t)

print("t-statistic:", t_stat)
print("Left-tailed p-value:", p_left)

t-statistic: 42.317706851929614
Left-tailed p-value: 1.0


### 3C. Calculate a two-tailed p-value

**Learning objective:** Calculate the probability for a difference in either direction.

$$H_0: \mu = 4 \qquad H_1: \mu \neq 4$$

$$p = 2P(T_{n-1} \geq |t_{\text{observed}}|)$$

In [29]:
# 3C. Calculate a two-tailed p-value
# Formula: p = 2 × P(T ≥ |t_observed|)

p_two = 2 * stats.t.sf(abs(t_stat), degrees_freedom)

print("t-statistic:", t_stat)
print("Two-tailed p-value:", p_two)

if p_two < 0.05:
    print("Reject H0: mean hospital stay differs from 4 days.")
else:
    print("Fail to reject H0.")

t-statistic: 42.317706851929614
Two-tailed p-value: 0.0
Reject H0: mean hospital stay differs from 4 days.


### 4A. Calculate a standardized score

**Learning objective:** Measure how far a hospital stay is from the mean.

The population formula is:

$$z = \frac{x-\mu}{\sigma}$$

Using our dataset's sample mean and sample standard deviation:

$$z = \frac{x-\bar{x}}{s}$$

Practice example: a hospital stay of 7 days.

In [30]:
# 4A. Calculate a standardized score
# Formula: z = (x - sample mean) / sample standard deviation

x = 7  # Practice hospital stay in days
z_score = (x - mean_stay) / std_stay

print("Hospital stay:", x, "days")
print("Standardized score:", round(z_score, 3))

Hospital stay: 7 days
Standardized score: 0.872


### 4B. Standardize a 2-day hospital stay

**Learning objective:** Interpret a negative standardized score.

$$z = \frac{x-\bar{x}}{s}
= \frac{2-4.396}{2.985}
\approx -0.803$$

In [31]:
# 4B. Standardize a 2-day hospital stay
# Formula: z = (x - sample mean) / sample standard deviation

x = 2
z_score = (x - mean_stay) / std_stay

print("Hospital stay:", x, "days")
print("Standardized score:", round(z_score, 3))

Hospital stay: 2 days
Standardized score: -0.803


### 4C. Calculate probabilities within standard deviations

**Learning objective:** Calculate the proportion within 1, 2, and 3
standard deviations for a normal distribution.

$$P(-k \leq Z \leq k) = \Phi(k)-\Phi(-k)$$

Here, $\Phi$ is the standard normal cumulative distribution function.

In [32]:
# 4C. Calculate probabilities within standard deviations
# Formula: P(-k ≤ Z ≤ k) = Φ(k) - Φ(-k)

for k in [1, 2, 3]:
    probability = stats.norm.cdf(k) - stats.norm.cdf(-k)
    print(f"Within {k} SD: {probability * 100:.2f}%")

Within 1 SD: 68.27%
Within 2 SD: 95.45%
Within 3 SD: 99.73%


### 4D. Calculate percentages within standard deviations

**Learning objective:** Compare actual hospital stays with the normal curve.

$$z_i = \frac{x_i-\bar{x}}{s}$$

$$\text{Percentage within }k\text{ SD}
= \frac{\text{Number of observations with }|z_i|\leq k}{n}\times100$$

In [33]:
# 4D. Calculate actual percentages in the diabetes dataset
# Formula: percentage = count(|z_i| ≤ k) / n × 100

z_stays = (hospital_stay - mean_stay) / std_stay

for k in [1, 2, 3]:
    count_within = (z_stays.abs() <= k).sum()
    actual_percent = count_within / n * 100
    normal_percent = (stats.norm.cdf(k) - stats.norm.cdf(-k)) * 100

    print(
        f"Within {k} SD: Diabetes = {actual_percent:.2f}%, "
        f"Normal = {normal_percent:.2f}%"
    )

Within 1 SD: Diabetes = 71.01%, Normal = 68.27%
Within 2 SD: Diabetes = 94.54%, Normal = 95.45%
Within 3 SD: Diabetes = 98.98%, Normal = 99.73%


### Topic 4 Summary: Standardized Scores

**Learning objective achieved:** Calculate and interpret standardized
scores, then compare observed percentages with a normal distribution.

$$z_i = \frac{x_i-\bar{x}}{s}$$

- A 7-day hospital stay has a score of 0.872: above the mean.
- A 2-day hospital stay has a score of -0.803: below the mean.
- A score of zero means the observation equals the mean.

| Range | Diabetes data | Normal distribution |
|---|---:|---:|
| Within 1 SD | 71.01% | 68.27% |
| Within 2 SD | 94.54% | 95.45% |
| Within 3 SD | 98.98% | 99.73% |

**Tools used:** `.abs()` removes signs, `<=` compares values,
`.sum()` counts True results, and `stats.norm.cdf()` calculates
standard normal cumulative probabilities.

**Conclusion:** Standardizing changes the scale, not the distribution's
shape. These percentages alone do not establish normality.

### 5A. Calculate an area using the standard normal distribution

**Learning objective:** Reproduce the instructor's Z-table value in Python.

$$P(0 \leq Z \leq 1.45)
= \Phi(1.45)-\Phi(0)
= \Phi(1.45)-0.5
\approx 0.4265$$

In [34]:
# 5A. Calculate the instructor's Z-table example
# Formula: area = Φ(1.45) - 0.5

z_instructor = 1.45
area = stats.norm.cdf(z_instructor) - 0.5

print("Area between 0 and 1.45:", round(area, 4))
print(f"Percentage: {area * 100:.2f}%")

Area between 0 and 1.45: 0.4265
Percentage: 42.65%


### 5B. Calculate the normal-curve area for a 7-day stay

**Learning objective:** Use a standardized hospital stay to calculate
an area under the standard normal curve.

$$z = \frac{7-\bar{x}}{s} \approx 0.872$$

$$P(0 \leq Z \leq z)=\Phi(z)-0.5$$

This is a normal-model practice calculation, not the actual percentage
of hospital stays in this range.

In [35]:
# 5B. Calculate the normal-curve area for a 7-day stay
# Formula: z = (7 - sample mean) / sample SD
z_seven = (7 - mean_stay) / std_stay

# Formula: area = Φ(z) - 0.5
area_seven = stats.norm.cdf(z_seven) - 0.5

print("Standardized score:", round(z_seven, 3))
print("Area from 0 to this score:", round(area_seven, 4))
print(f"Normal-model percentage: {area_seven * 100:.2f}%")

Standardized score: 0.872
Area from 0 to this score: 0.3085
Normal-model percentage: 30.85%


### 5C. Calculate the right-tail area

**Learning objective:** Calculate the normal-curve area above a score.

$$P(Z \geq z)=1-\Phi(z)$$

Instructor example:
$$P(Z \geq 1.45)\approx 0.0735$$

Diabetes practice score:
$$P(Z \geq 0.872)\approx 0.1915$$

These are normal-model areas. The individual hospital-stay score
is not a hypothesis test of the population mean.

In [36]:
# 5C. Calculate the right-tail area
# Formula: right-tail area = 1 - Φ(z)
# stats.norm.sf(z) calculates this directly.

right_instructor = stats.norm.sf(z_instructor)
right_seven = stats.norm.sf(z_seven)

print(f"Instructor right-tail area: {right_instructor:.4f}")
print(f"7-day score right-tail area: {right_seven:.4f}")

Instructor right-tail area: 0.0735
7-day score right-tail area: 0.1915


### 5D. Calculate a critical Z-value

**Learning objective:** Find the cutoff leaving 5% in the right tail.

$$\alpha = 0.05$$

$$z_{\text{critical}}=\Phi^{-1}(1-\alpha)
=\Phi^{-1}(0.95)\approx 1.645$$

For a valid right-tailed Z-test, reject the null hypothesis when
the observed test statistic exceeds this cutoff.

In [37]:
# 5D. Calculate a critical Z-value
# Formula: z_critical = Φ⁻¹(1 - alpha)

alpha = 0.05
z_critical = stats.norm.ppf(1 - alpha)

print("Critical Z-value:", round(z_critical, 3))
print("Instructor score:", z_instructor)
print("Instructor score exceeds cutoff:", z_instructor > z_critical)

Critical Z-value: 1.645
Instructor score: 1.45
Instructor score exceeds cutoff: False


### 5E. Calculate two-tailed critical Z-values

**Learning objective:** Find the cutoffs for a two-tailed Z-test.

$$\alpha=0.05 \qquad \frac{\alpha}{2}=0.025$$

$$z_{\text{lower}}=\Phi^{-1}(0.025)\approx -1.960$$

$$z_{\text{upper}}=\Phi^{-1}(0.975)\approx 1.960$$

For a valid two-tailed Z-test, reject the null hypothesis when:

$$|z_{\text{test}}|>1.960$$

In [38]:
# 5E. Calculate two-tailed critical Z-values
alpha = 0.05

# Formula: lower cutoff = Φ⁻¹(alpha / 2)
z_lower = stats.norm.ppf(alpha / 2)

# Formula: upper cutoff = Φ⁻¹(1 - alpha / 2)
z_upper = stats.norm.ppf(1 - alpha / 2)

print("Lower critical Z:", round(z_lower, 3))
print("Upper critical Z:", round(z_upper, 3))

Lower critical Z: -1.96
Upper critical Z: 1.96


In [39]:
# 5F. Calculate the critical t-value
# Objective: Compare your diabetes t-statistic with the cutoff.
# Formula: t_critical = inverse t CDF(1 - alpha / 2, df)

alpha = 0.05
t_critical = stats.t.ppf(1 - alpha / 2, degrees_freedom)

print("Critical t-value:", round(t_critical, 4))
print("Observed t-statistic:", round(t_stat, 4))

if abs(t_stat) > t_critical:
    print("Reject H0: mean hospital stay differs from 4 days.")
else:
    print("Fail to reject H0.")

Critical t-value: 1.96
Observed t-statistic: 42.3177
Reject H0: mean hospital stay differs from 4 days.


### Topic 5 Summary: Curve Areas and Critical Values

**Learning objectives achieved:** Calculate normal-curve areas,
find critical values, and apply the critical-value method.

| Calculation | Python tool | Result |
|---|---|---|
| Instructor area between 0 and 1.45 | `stats.norm.cdf(1.45) - 0.5` | 0.4265 |
| Instructor right-tail area | `stats.norm.sf(1.45)` | 0.0735 |
| 7-day score right-tail area, normal model | `stats.norm.sf(z_seven)` | 0.1915 |
| Right-tailed Z-cutoff at alpha = 0.05 | `stats.norm.ppf(0.95)` | 1.645 |
| Two-tailed Z-cutoffs at alpha = 0.05 | `stats.norm.ppf()` | ±1.960 |
| Diabetes two-tailed t-cutoffs | `stats.t.ppf()` | Approximately ±1.960 |

**Diabetes mean-test decision:**

$$|t_{\text{observed}}|=42.3177>t_{\text{critical}}\approx1.9600$$

Reject the null hypothesis that the mean hospital stay equals our
4-day practice benchmark. The sample mean is approximately 4.396 days.

Both the p-value method and critical-value method give the same decision.

**Remember:** `cdf()` gives left-side area, `sf()` gives right-side area,
and `ppf()` converts a cumulative probability into a cutoff.

The inference assumes independent observations; repeated admissions
may violate this assumption. Statistical significance alone does not
establish practical importance.

### 6A. Run a one-sample t-test

**Learning objective:** Use SciPy to test whether the mean hospital
stay differs from our 4-day practice benchmark.

$$H_0:\mu=4 \qquad H_1:\mu\neq4$$

$$t=\frac{\bar{x}-\mu_0}{s/\sqrt{n}}$$

The instructor's formula compares the sample mean with an assumed mean.
Here, the assumed mean is 4 days.

In [40]:
# 6A. Run a one-sample t-test
# Formula: t = (sample mean - assumed mean) / (s / sqrt(n))

test_result = stats.ttest_1samp(
    hospital_stay,
    popmean=4,
    alternative="two-sided"
)

print("SciPy t-statistic:", test_result.statistic)
print("Manual t-statistic:", t_stat)
print("SciPy p-value:", test_result.pvalue)

SciPy t-statistic: 42.317706851929614
Manual t-statistic: 42.317706851929614
SciPy p-value: 0.0


### 6B. Interpret the one-sample t-test

**Learning objective:** Convert the p-value into a test decision.

$$p<\alpha \Rightarrow \text{Reject }H_0$$

$$p\geq\alpha \Rightarrow \text{Fail to reject }H_0$$

Here, $\alpha=0.05$ and our practice null hypothesis is $\mu=4$ days.

In [41]:
# 6B. Interpret the one-sample t-test
# Decision rule: reject H0 when p < alpha

alpha = 0.05

print(f"Sample mean: {mean_stay:.3f} days")
print(f"t-statistic: {test_result.statistic:.4f}")

if test_result.pvalue < 0.001:
    print("p < .001")
else:
    print(f"p = {test_result.pvalue:.4f}")

if test_result.pvalue < alpha:
    print("Reject H0: the mean hospital stay differs from 4 days.")
else:
    print("Fail to reject H0.")

Sample mean: 4.396 days
t-statistic: 42.3177
p < .001
Reject H0: the mean hospital stay differs from 4 days.


### Topic 6 Summary: One-Sample t-Test

**Learning objective achieved:** Run and interpret a one-sample t-test,
and compare its result with the manual calculation.

**Question:** Does the mean hospital stay differ from 4 days?
Four days is our practice benchmark.

$$H_0:\mu=4 \qquad H_1:\mu\neq4$$

$$t=\frac{\bar{x}-\mu_0}{s/\sqrt{n}}$$

| Result | Value |
|---|---:|
| Number of hospital encounters | 101,766 |
| Sample mean | 4.396 days |
| Sample standard deviation | 2.985 days |
| Degrees of freedom | 101,765 |
| t-statistic | 42.3177 |
| p-value | < .001 |
| Significance level | .05 |
| Decision | Reject H0 |

**Tool:** `stats.ttest_1samp()` compares one sample mean with a benchmark.
`popmean=4` sets the benchmark; `alternative="two-sided"` tests either direction.

**Conclusion:** Under the test's assumptions, there is evidence that
the mean hospital stay differs from 4 days. The estimated difference
is 0.396 days.

Repeated admissions may violate independence. Statistical significance
does not establish practical importance.

### 7A. Calculate the instructor's t-table cutoff

**Learning objective:** Select a t-table cutoff using sample size
and significance level.

Instructor example:

$$n=25 \qquad df=n-1=24$$

$$\alpha=0.05$$

$$t_{\text{critical}}=F^{-1}_{t,24}(0.95)\approx1.711$$

This same positive cutoff applies to a two-tailed test with alpha = 0.10.

In [42]:
# 7A. Calculate the instructor's t-table cutoff
# These values come from the instructor's example.

n_instructor = 25
df_instructor = n_instructor - 1  # Formula: df = n - 1
alpha_instructor = 0.05

# Formula: t_critical = inverse t CDF(1 - alpha, df)
cutoff_instructor = stats.t.ppf(
    1 - alpha_instructor, df_instructor
)

print("Instructor degrees of freedom:", df_instructor)
print("Instructor one-tailed cutoff:", round(cutoff_instructor, 3))

Instructor degrees of freedom: 24
Instructor one-tailed cutoff: 1.711


### 7B. Calculate the diabetes t-cutoff

**Learning objective:** Compare t-cutoffs for different sample sizes,
using a right-tailed test with alpha = 0.05.

$$df = n - 1$$

$$t_{\text{critical}} = F^{-1}_{t,\,df}(1-\alpha)$$

For our diabetes data:

$$df = 101766 - 1 = 101765$$

In [43]:
# 7B. Calculate the diabetes t-cutoff
n_diabetes = hospital_stay.count()
df_diabetes = n_diabetes - 1       # Formula: df = n - 1
alpha = 0.05

# Formula: t_critical = inverse t CDF(1 - alpha, df)
cutoff_diabetes = stats.t.ppf(1 - alpha, df_diabetes)

print("Diabetes sample size:", n_diabetes)
print("Diabetes degrees of freedom:", df_diabetes)
print("Instructor cutoff:", round(cutoff_instructor, 3))
print("Diabetes cutoff:", round(cutoff_diabetes, 3))

Diabetes sample size: 101766
Diabetes degrees of freedom: 101765
Instructor cutoff: 1.711
Diabetes cutoff: 1.645


### 8A. Calculate the risk of at least one false positive

**Learning objective:** Understand why performing many hypothesis
tests increases false-positive risk.

For m independent tests, when all null hypotheses are true:

$$P(\text{at least one false positive}) = 1-(1-\alpha)^m$$

Practice example: Suppose we conduct 10 independent tests
about diabetes hospital stays, each at alpha = 0.05.

$$1-(1-0.05)^{10} \approx 0.4013$$

These are hypothetical practice assumptions, not results
calculated from our diabetes dataset.

In [44]:
# 8A. Calculate the risk of at least one false positive
alpha = 0.05
m = 10  # Assumed number of independent tests for practice

# Formula: P(at least one false positive) = 1 - (1 - alpha)^m
false_positive_risk = 1 - (1 - alpha) ** m

print(f"False-positive risk: {false_positive_risk:.2%}")

False-positive risk: 40.13%


### 8B. Calculate the Bonferroni threshold

**Learning objective:** Adjust the significance threshold when
testing multiple hypotheses.

$$\alpha_{\text{adjusted}} = \frac{\alpha}{m}$$

For our hypothetical family of 10 diabetes-related tests:

$$\alpha_{\text{adjusted}} = \frac{0.05}{10} = 0.005$$

Reject a null hypothesis when its p-value is below 0.005.
This controls the probability of at least one false positive
across the family at no more than 5%, assuming valid p-values.

In [45]:
# 8B. Calculate the Bonferroni threshold
alpha = 0.05
m = 10  # Hypothetical number of tests

alpha_adjusted = alpha / m  # Formula: adjusted alpha = alpha / m

print("Adjusted threshold:", alpha_adjusted)
print(f"Adjusted threshold as percentage: {alpha_adjusted:.2%}")

Adjusted threshold: 0.005
Adjusted threshold as percentage: 0.50%


### 8C. Make a decision using the Bonferroni threshold

**Learning objective:** Compare a diabetes test p-value with
an adjusted significance threshold.

$$H_0:\mu=4 \qquad H_1:\mu\ne4$$

$$\text{Reject }H_0\text{ if }p<\frac{\alpha}{m}$$

We use our actual hospital-stay data and a practice benchmark
of 4 days. For illustration, suppose this test belongs to a
planned family of 10 tests:

$$\alpha_{\text{adjusted}}=\frac{0.05}{10}=0.005$$

We have not performed all 10 tests; this demonstrates the decision rule.

In [46]:
# 8C. Apply Bonferroni to the diabetes test
result_diabetes = stats.ttest_1samp(
    hospital_stay,
    popmean=4,
    alternative="two-sided"
)

alpha_adjusted = 0.05 / 10  # Formula: adjusted alpha = alpha / m

print(f"t-statistic: {result_diabetes.statistic:.3f}")
print("p < 0.001" if result_diabetes.pvalue < 0.001
      else f"p = {result_diabetes.pvalue:.4f}")

# Decision rule: reject H0 if p < adjusted alpha
if result_diabetes.pvalue < alpha_adjusted:
    print("Reject H0 at the adjusted threshold of 0.005.")
else:
    print("Fail to reject H0 at the adjusted threshold of 0.005.")

t-statistic: 42.318
p < 0.001
Reject H0 at the adjusted threshold of 0.005.


### 8D. Calculate expected false positives

**Learning objective:** Calculate how many false positives
to expect when all tested null hypotheses are true.

$$E(\text{false positives}) = m\alpha$$

Instructor example: 20 color tests at alpha = 0.05.

$$20(0.05)=1$$

Diabetes practice example: 10 tests at alpha = 0.05.

$$10(0.05)=0.5$$

These are averages across repeated studies, not guaranteed counts.
The diabetes test count is hypothetical.

In [47]:
# 8D. Calculate expected false positives
alpha = 0.05

m_instructor = 20  # Color tests in the instructor's comic
m_diabetes = 10    # Hypothetical diabetes tests

# Formula: expected false positives = m * alpha
expected_instructor = m_instructor * alpha
expected_diabetes = m_diabetes * alpha

print("Instructor expected false positives:", expected_instructor)
print("Diabetes practice expected false positives:", expected_diabetes)

Instructor expected false positives: 1.0
Diabetes practice expected false positives: 0.5


### 9A. Calculate the observed difference in mean hospital stay

**Learning objective:** Calculate the statistic we will use
in a permutation test.

The instructor uses two groups, A and B.
For our diabetes data, we use Female and Male encounters.

$$T_{\text{observed}}=\bar{x}_{Female}-\bar{x}_{Male}$$

$$H_0:\text{Hospital-stay distributions are identical across groups}$$

Under the null and exchangeability assumption, we can shuffle
group labels while preserving group sizes.

In [48]:
# 9A. Calculate the observed difference using your dataset
female_stay = df.loc[
    df["gender"] == "Female", "time_in_hospital"
].dropna()

male_stay = df.loc[
    df["gender"] == "Male", "time_in_hospital"
].dropna()

# Formula: observed difference = female mean - male mean
observed_difference = female_stay.mean() - male_stay.mean()

print("Female sample size:", female_stay.count())
print("Male sample size:", male_stay.count())
print(f"Female mean stay: {female_stay.mean():.3f} days")
print(f"Male mean stay: {male_stay.mean():.3f} days")
print(f"Observed difference: {observed_difference:.3f} days")

Female sample size: 54708
Male sample size: 47055
Female mean stay: 4.479 days
Male mean stay: 4.300 days
Observed difference: 0.179 days


### 9B. Shuffle the hospital stays into two groups

**Learning objective:** Generate one mean difference under
the permutation-test null hypothesis.

Pool the stays, shuffle them, and split them into groups
with the original sample sizes.

$$T_{\text{shuffled}}=\bar{x}_{A^*}-\bar{x}_{B^*}$$

The shuffled groups are artificial assignments.
This is equivalent to shuffling group labels.

For this practice test, we assume encounters are independent
and labels are exchangeable under the null. Repeat admissions
can violate this assumption.

In [49]:
# 9B. Perform one permutation using your diabetes data
rng = np.random.default_rng(42)  # Seed makes practice reproducible

combined_stays = np.concatenate([
    female_stay.to_numpy(),
    male_stay.to_numpy()
])

n_female = len(female_stay)
shuffled_stays = rng.permutation(combined_stays)

# Preserve the original group sizes
shuffled_female = shuffled_stays[:n_female]
shuffled_male = shuffled_stays[n_female:]

# Formula: T_shuffled = mean(A*) - mean(B*)
shuffled_difference = (
    shuffled_female.mean() - shuffled_male.mean()
)

print(f"Observed difference: {observed_difference:.3f} days")
print(f"One shuffled difference: {shuffled_difference:.3f} days")

Observed difference: 0.179 days
One shuffled difference: -0.000 days


### 9C. Calculate the permutation p-value

**Learning objective:** Estimate how often shuffled differences
are at least as large in magnitude as our observed difference.

$$T_b=\bar{x}_{A_b^*}-\bar{x}_{B_b^*}$$

$$k=\#\{b:|T_b|\geq|T_{\text{observed}}|\}$$

$$\hat{p}=\frac{k+1}{B+1}$$

B is the number of random shuffles.
k counts equally or more extreme shuffled differences.
Adding 1 prevents reporting a zero Monte Carlo p-value.



In [50]:
# 9C. Repeat the permutation using your diabetes data
rng = np.random.default_rng(42)
B = 999
permutation_differences = np.empty(B)

for b in range(B):
    shuffled = rng.permutation(combined_stays)

    # Formula: T_b = mean(A*) - mean(B*)
    permutation_differences[b] = (
        shuffled[:n_female].mean()
        - shuffled[n_female:].mean()
    )

# Count extreme differences in either direction
k = np.count_nonzero(
    np.abs(permutation_differences) >= abs(observed_difference)
)

p_permutation = (k + 1) / (B + 1)  # Formula: p = (k + 1)/(B + 1)

print("Number of shuffles:", B)
print("Extreme shuffled differences:", k)
print(f"Estimated permutation p-value: {p_permutation:.4f}")
print("Reject H0." if p_permutation < 0.05
      else "Fail to reject H0.")

Number of shuffles: 999
Extreme shuffled differences: 0
Estimated permutation p-value: 0.0010
Reject H0.


### 10A. Calculate the confidence level

**Learning objective:** Convert alpha into a confidence level.

$$\text{Confidence level}=1-\alpha$$

Instructor example:

$$1-0.03=0.97=97\%$$

For our diabetes analysis, we choose alpha = 0.05:

$$1-0.05=0.95=95\%$$

A 95% confidence interval procedure captures the true population
parameter in about 95% of repeated samples, under its assumptions.
It is not a range containing 95% of individual hospital stays.

In [51]:
# 10A. Calculate confidence levels
alpha_instructor_ci = 0.03
alpha_diabetes_ci = 0.05

# Formula: confidence level = 1 - alpha
confidence_instructor = 1 - alpha_instructor_ci
confidence_diabetes = 1 - alpha_diabetes_ci

print(f"Instructor confidence level: {confidence_instructor:.0%}")
print(f"Diabetes confidence level: {confidence_diabetes:.0%}")

Instructor confidence level: 97%
Diabetes confidence level: 95%


### 10B. Calculate the confidence interval for the mean

**Learning objective:** Calculate the margin of error and
confidence limits using our diabetes data.

$$SE=\frac{s}{\sqrt{n}}$$

$$t^*=F^{-1}_{t,n-1}(1-\alpha/2)$$

$$ME=t^*\times SE$$

$$CI=[\bar{x}-ME,\;\bar{x}+ME]$$

For a 95% interval, alpha = 0.05.
We divide alpha by 2 because the interval has two tails.

Using our diabetes results:

$$CI\approx4.39599\pm1.9600(0.0093575)$$

In [52]:
# 10B. Calculate the 95% CI using your diabetes data
n_ci = hospital_stay.count()
mean_ci = hospital_stay.mean()
s_ci = hospital_stay.std(ddof=1)

se_ci = s_ci / np.sqrt(n_ci)  # Formula: SE = s / sqrt(n)
alpha_ci = 0.05

# Formula: t* = inverse t CDF(1 - alpha/2, n - 1)
t_star = stats.t.ppf(1 - alpha_ci / 2, n_ci - 1)

margin_error = t_star * se_ci  # Formula: ME = t* × SE
ci_lower = mean_ci - margin_error
ci_upper = mean_ci + margin_error

print(f"Mean stay: {mean_ci:.4f} days")
print(f"Margin of error: {margin_error:.4f} days")
print(f"95% confidence interval: [{ci_lower:.4f}, {ci_upper:.4f}] days")

Mean stay: 4.3960 days
Margin of error: 0.0183 days
95% confidence interval: [4.3776, 4.4143] days


### 10C. Calculate the interval midpoint and margin of error

**Learning objective:** Convert an interval into midpoint ± half-width.

$$\text{Midpoint}=\frac{L+U}{2}$$

$$\text{Half-width}=\frac{U-L}{2}$$

Instructor interval: [2.5, 5.3]

$$\text{Midpoint}=3.9,\qquad \text{Half-width}=1.4$$

Our symmetric diabetes t-interval:

$$4.3960\pm0.0183\text{ days}$$

Here, the midpoint is the sample mean and the half-width
is the margin of error.

Interpretation: The interval estimates the population mean.
The 95% confidence level describes the method's long-run coverage,
under its assumptions.

In [53]:
# 10C. Calculate midpoint and half-width
lower_instructor = 2.5
upper_instructor = 5.3

# Formula: midpoint = (L + U) / 2
midpoint_instructor = (lower_instructor + upper_instructor) / 2
midpoint_diabetes = (ci_lower + ci_upper) / 2

# Formula: half-width = (U - L) / 2
half_width_instructor = (upper_instructor - lower_instructor) / 2
half_width_diabetes = (ci_upper - ci_lower) / 2

print(f"Instructor: {midpoint_instructor:.1f} ± {half_width_instructor:.1f}")
print(f"Diabetes: {midpoint_diabetes:.4f} ± {half_width_diabetes:.4f} days")

Instructor: 3.9 ± 1.4
Diabetes: 4.3960 ± 0.0183 days


### 11A. Practice the known-variance confidence interval

**Learning objective:** Code the z-interval when population
standard deviation is known.

$$SE=\frac{\sigma}{\sqrt{n}}$$

$$z^*=\Phi^{-1}(1-\alpha/2)$$

$$CI=\bar{x}\pm z^*\frac{\sigma}{\sqrt{n}}$$

Use our actual diabetes mean and sample size.
For this exercise only, assume the population standard
deviation is known to be 3 days.

$$CI=4.39599\pm1.95996\frac{3}{\sqrt{101766}}$$

The value sigma = 3 is a practice assumption, not a dataset result.

In [54]:
# 11A. Practice the instructor's known-variance formula
sigma_practice = 3.0  # Assumed known population SD, in days
alpha_z = 0.05

# Formula: z* = inverse normal CDF(1 - alpha/2)
z_star = stats.norm.ppf(1 - alpha_z / 2)

se_known = sigma_practice / np.sqrt(n_ci)  # SE = sigma / sqrt(n)
margin_z = z_star * se_known              # ME = z* × SE

lower_z = mean_ci - margin_z
upper_z = mean_ci + margin_z

print(f"z critical value: {z_star:.4f}")
print(f"Practice margin of error: {margin_z:.4f} days")
print(f"Practice 95% z-interval: [{lower_z:.4f}, {upper_z:.4f}] days")

z critical value: 1.9600
Practice margin of error: 0.0184 days
Practice 95% z-interval: [4.3776, 4.4144] days


### 12A. Confidence interval with unknown population variance

**Learning objective:** Use the sample standard deviation
and verify our manual t-interval with SciPy.

$$s^2=\frac{\sum_{i=1}^{n}(x_i-\bar{x})^2}{n-1}$$

$$CI=\bar{x}\pm t_{1-\alpha/2,n-1}\frac{s}{\sqrt{n}}$$

Our diabetes population variance is unknown.
We estimate it from the observed hospital stays.

For independent observations, the t-interval is exact for
a normal population and approximately valid for large samples
under suitable conditions.

In [55]:
# 12A. Verify the manual t-interval using SciPy
# Formula: CI = mean ± t* × s / sqrt(n)
ci_scipy = stats.t.interval(
    confidence=0.95,
    df=n_ci - 1,
    loc=mean_ci,
    scale=se_ci
)

print(f"Sample variance: {hospital_stay.var(ddof=1):.4f}")
print(f"Sample standard deviation: {s_ci:.4f} days")
print(f"SciPy 95% CI: [{ci_scipy[0]:.4f}, {ci_scipy[1]:.4f}] days")
print("Matches manual interval:",
      np.allclose(ci_scipy, [ci_lower, ci_upper]))

Sample variance: 8.9109
Sample standard deviation: 2.9851 days
SciPy 95% CI: [4.3776, 4.4143] days
Matches manual interval: True


### 12B. Compare 95% t critical values

**Learning objective:** Calculate the t critical value using
sample size and confidence level.

$$df=n-1$$

$$t^*=F^{-1}_{t,df}(1-\alpha/2)$$

Instructor example: n = 19, alpha = 0.05.

$$df=19-1=18,\qquad t^*\approx2.101$$

Diabetes data: n = 101766.

$$df=101765,\qquad t^*\approx1.960$$

In [56]:
# 12B. Calculate 95% t critical values
n_slide = 19
alpha_compare = 0.05

df_slide = n_slide - 1  # Formula: df = n - 1

# Formula: t* = inverse t CDF(1 - alpha/2, df)
t_slide = stats.t.ppf(1 - alpha_compare / 2, df_slide)
t_diabetes = stats.t.ppf(1 - alpha_compare / 2, n_ci - 1)

print("Instructor degrees of freedom:", df_slide)
print(f"Instructor t critical value: {t_slide:.3f}")
print(f"Diabetes t critical value: {t_diabetes:.3f}")

Instructor degrees of freedom: 18
Instructor t critical value: 2.101
Diabetes t critical value: 1.960


### 13A. Calculate mean model accuracy

**Learning objective:** Summarize the instructor's model scores
before hypothesis testing.

$$\bar{x}_A=\frac{1}{12}\sum_{i=1}^{12}A_i$$

$$\bar{x}_B=\frac{1}{12}\sum_{i=1}^{12}B_i$$

$$\text{Mean difference}=\bar{x}_A-\bar{x}_B$$

These accuracy scores come from the instructor's slide.
For our diabetes hospital-stay regression models, we will
compare an appropriate metric such as MAE.

In [57]:
# 13A. Enter the instructor's 12-fold accuracy scores
model_a = np.array([
    0.943926, 0.943017, 0.944070, 0.945392,
    0.944855, 0.942191, 0.942852, 0.944793,
    0.943450, 0.943160, 0.940290, 0.943470
])

model_b = np.array([
    0.656347, 0.651989, 0.641827, 0.652567,
    0.648726, 0.649779, 0.650523, 0.653662,
    0.655080, 0.657434, 0.657414, 0.652064
])

# Formula: mean = sum of scores / number of scores
mean_a = model_a.mean()
mean_b = model_b.mean()
accuracy_difference = mean_a - mean_b

print(f"Model A mean accuracy: {mean_a:.2%}")
print(f"Model B mean accuracy: {mean_b:.2%}")
print(f"Difference: {accuracy_difference * 100:.2f} percentage points")

Model A mean accuracy: 94.35%
Model B mean accuracy: 65.23%
Difference: 29.12 percentage points


### 13B. Paired t-test — instructor example

$$H_0:\mu_d=0 \qquad H_1:\mu_d\ne0$$

$$d_i=A_i-B_i$$

$$t=\frac{\bar d}{s_d/\sqrt{n}},\qquad df=n-1$$

In [58]:
differences = model_a - model_b

n_pairs = len(differences)
mean_difference = differences.mean()
sd_difference = differences.std(ddof=1)

# Formula: t = mean difference / standard error
standard_error = sd_difference / np.sqrt(n_pairs)
t_statistic = mean_difference / standard_error

df_pairs = n_pairs - 1
p_paired = 2 * stats.t.sf(abs(t_statistic), df=df_pairs)

print(f"t-statistic: {t_statistic:.2f}")
print(f"Degrees of freedom: {df_pairs}")
print(f"p-value: {p_paired:.3e}")
print("Reject H₀" if p_paired < 0.05 else "Fail to reject H₀")

t-statistic: 202.81
Degrees of freedom: 11
p-value: 5.253e-21
Reject H₀


### 13C. Compare one-sided and two-sided tests

Let each paired difference be $d_i=A_i-B_i$.

$$t=\frac{\bar d}{s_d/\sqrt{n}},\qquad df=n-1$$

Two-sided:
$$H_0:\mu_d=0 \qquad H_1:\mu_d\ne0$$

Right-sided:
$$H_0:\mu_d\le0 \qquad H_1:\mu_d>0$$

Choose the direction before examining the results.

In [59]:
# Instructor's second example—not diabetes results
a_close = np.array([
    .661420, .656604, .650470, .666336,
    .685282, .684663, .677166, .647430,
    .639121, .649021, .641169, .684656
])

b_close = np.array([
    .649655, .651948, .650936, .646970,
    .650646, .652980, .654323, .657978,
    .655885, .652126, .647335, .655204
])

# Formula: t = mean paired difference / standard error
d_close = a_close - b_close
n_close = len(d_close)
se_close = d_close.std(ddof=1) / np.sqrt(n_close)
t_close = d_close.mean() / se_close
df_close = n_close - 1

p_two_close = 2 * stats.t.sf(abs(t_close), df=df_close)
p_right_close = stats.t.sf(t_close, df=df_close)

print(f"t = {t_close:.3f}; df = {df_close}")
print(f"Two-sided p = {p_two_close:.4f}")
print(f"Right-sided p = {p_right_close:.4f}")

for label, p in [
    ("Two-sided", p_two_close),
    ("Right-sided", p_right_close)
]:
    decision = "Reject H₀" if p < 0.05 else "Fail to reject H₀"
    print(f"{label}: {decision}")

t = 1.921; df = 11
Two-sided p = 0.0810
Right-sided p = 0.0405
Two-sided: Fail to reject H₀
Right-sided: Reject H₀
